### PR TITLE
Implement docstrings

### PR DESCRIPTION
**Pull request by arg0d**
Docstrings are declared in .udl file by prefixing a line with three slashes ///. Docstrings can be placed basically anywhere - functions, objects, methods, constructors, callbacks, etc.. A docstring placed in a wrong place will generate UDL parser error.

### CODE REVIEW

**Comment by arg0d** [comment]
This PR implements backend for parsing docstrings decsribed in https://github.com/mozilla/uniffi-rs/issues/200#issuecomment-1469934813

**Review by badboy — COMMENTED** [review]
cc @mhammond / @bendk -- would like to hear your general feeling about this feature (I'll take care of the code review).

I think the feature itself is ok. Glean re-exports some methods directly and I bet having some docs on them would be useful.
We cannot make use of language-specific syntax of course (e.g. javadoc-style vs swift-doc style parameter annotation), but it's something.  

IMO the force-use of triple-slashes (`///`) within the doc attribute feels a bit forced.
The attribute supports multi-line strings regardless.
Obviously just normal comments in the UDL would be the more perfect solution, but we won't get that unless we modify weedle2 significantly.
I'd be ok to just take the string as is.

The alternative is parsing Rust comments (see [this approach](https://github.com/mozilla/uniffi-rs/issues/200#issuecomment-1477893662)), but this requires parsing the Rust code as well.
Has the upside of then being what the macro code will have eventually anyway.

@arg0d I already left some review comments, but please hold off until the others can chime in so you can avoid doing potentially unnecessary work.

**Comment by mhammond** [comment]
> would like to hear your general feeling about this feature

I agree with everything you wrote :) I agree the forced `///` seems a bit, um, forced, but that's a matter of taste, and I guess it does make it easier to paste them into rust code later - which I think is the future here anyway - ie, I see this as a somewhat temporary solution while we migrate to procmacros, so don't feel particularly invested in stylistic issues.

File: uniffi_bindgen/src/interface/attributes.rs
Comment (bendk) [inline uniffi_bindgen/src/interface/attributes.rs:115]: I always appreciate comments like this -- one less trip to the regex docs.

**Review by bendk — COMMENTED** [review]
I'll let @badboy handle the code review, I just skimmed it.  On the general approach, I agree 100% with @mhammond:

> I see this as a somewhat temporary solution while we migrate to procmacros, so don't feel particularly invested in stylistic issues.

If the plan was to stay with UDL forever, then I would consider a different approach, like implementing a parser that handled comments.  But if we're going to be moving to proc-macros -- and I'm pretty sure we are -- then this seems ilke a better path.

**Comment by badboy** [comment]
Thank you to both, I'll finish up the code review then and we can get this into a mergable state.

**Comment by zduny** [comment]
@arg0d Yes, we do plan to create a PR - I was simply checking first if you're interested - it may take some time though - we desynced slightly from upstream, I'm working now on rebasing - expect a PR tomorrow or in the beginning of the next week.

**Comment by badboy** [comment]
Then I hold of on a further review here. Having that directly as comments in the UDL would definitely be nicer than the doc attribute.

**Comment by zduny** [comment]
@arg0d I've created a [PR](https://github.com/mozilla/uniffi-rs/pull/1498) with our documentation solution.

Sadly, we can't rework it to use comments in `.udl` files - I have no experience with `nom` parsers, also we have no capacity to do so at the moment.

**Comment by mhammond** [comment]
> @zduny would you consider reworking your PR to take docstring input from .udl instead of parsing Ruts code?

As @bendk and I wrote above, I think having Rust docstrings be canonical is actually a much better idea, mainly as there's a sane path forward for procmacros.

So @arg0d, how do you feel about just taking the approach in #1498? If you really don't like that idea, then I think we must at least integrate the 2 approaches - eg, having the UDL parser extract the comments, but using the exact same crates and templates in that PR. Note I'm not suggesting that the other PR has exactly the correct setup there, the point is that we don't want multiple "back ends" for docstrings.

**Comment by arg0d** [comment]
@zduny I have implemented parsing docstring in .udl. You can see the changes in this PR. To me it looks like we could easily combine both of these PRs. The only concern I would have with combining them, is which source of docstrings is superior: Rust or UDL. If docstrings can be defined in both Rust and UDL, then which version should be used if a docstring is defined in both places? To me it looks like UDL docstring should be the higher priority, while Rust docstring should act as a backup if UDL docstring is not available. Parsing docstrings from UDL files is the more natural method, because UDL files are the current standard API for working with uniffi.

@mhammond I agree that having duplicate docstring logic in bindings generators is undesirable. However, since the current standard API of uniffi are UDL files, I think parsing docstrings from UDL files is superior to parsing docstrings from Rust code. Forcing users to write docstrings in Rust code, while proc macros aren't finished yet, does not seem like a good user experience to me. When proc macros become the standard API in uniffi, and UDL files aren't necessary anymore, I would be glad write docstrings in Rust code. Until then, I think writing docstrings in UDL files is the more natural method.

**Comment by arg0d** [comment]
@zduny I have implemented parsing docstring in .udl. You can see the changes in this PR. To me it looks like we could easily combine both of these PRs. The only concern I would have with combining them, is which source of docstrings is superior: Rust or UDL. If docstrings can be defined in both Rust and UDL, then which version should be used if a docstring is defined in both places? To me it looks like UDL docstring should be the higher priority, while Rust docstring should act as a backup if UDL docstring is not available. Parsing docstrings from UDL files is the more natural method, because UDL files are the current standard API for working with uniffi.

@mhammond I agree that having duplicate docstring logic in bindings generators is undesirable. However, since the current standard API of uniffi are UDL files, I think parsing docstrings from UDL files is superior to parsing docstrings from Rust code. Forcing users to write docstrings in Rust code, while proc macros aren't finished yet, does not seem like a good user experience to me. When proc macros become the standard API in uniffi, and UDL files aren't necessary anymore, I would be glad write docstrings in Rust code. Until then, I think writing docstrings in UDL files is the more natural method.

**Comment by mhammond** [comment]
>  I think parsing docstrings from UDL files is superior to parsing docstrings from Rust code. Forcing users to write docstrings in Rust code

Our experience in app-services is that even while we use UDL, docstrings in rust code is superior - that's where they have naturally been written by the authors of the component. We find that having the docstrings be so far apart from the implementation means they aren't added to the UDL.

However, that's not really my concern - I don't really mind allowing them to be added to both. But it raises the question: how are these docstrings actually going to be consumed? Is the intention that rustdoc ends up being used to create a "book" which is used by all foreign consumers? Or should the focus be on "round-tripping" those docstrings into the foreign bindings to better support things like autocomplete in IDEs etc? Or both? The latter implies more structure of the strings might be necessary. #1498 appears to take the latter approach, which has a lot of appeal to me.

So that's where I think a common approach is important - *where* the docstrings gets written matters less to me than how these docstrings actually end up being used. In more concrete terms, this patch having documentation be a simple `str` where that other patch uses a more structured `uniffi_docs::Structure` is something that needs to be thought through in the context of how the docs are going to actually be used/consumed, and one approach chosen over the other.

**Comment by arg0d** [comment]
You raised a very interesting proposition of the docbook, which I haven't thought of before. It would be external documentation, something like [uniffi-rs docs](https://mozilla.github.io/uniffi-rs/), right? I'm gonna discuss this idea with my colleagues later this week.

The current "user story" we had in mind is the following: **make it easier for bindings consumers to integrate libraries by providing some form of documentation, not just code**. In that sense, I think having both IDE docstrings and an external docbook could be complementary. Even if the content of both of these forms were the same, providing documentation in multiple formats can be useful for different use cases. For example, IDE docstrings are useful when writing/reading the code, because the IDE can show documentation that is relevant to the code being displayed on screen, such as function descriptions. An example for a docbook could be a developer who is new to the library and wants to get a general overview of functionality that is available. In that case, concentrated documentation in the form of a docbook is better suited than having to comb through heaps of source code to find nuggets of documentation.

Last week, I had a conversation with my colleague @nymoral about what he would expect for documentation to do. The general direction of discussion was to provide docstrings that are properly displayed by IDEs for all supported languages. We tried to come up with a list of requirements that docstrings should support.
- general description of a function
- parameters of a function
- return value of a function
- error values of a function (errors/exceptions)
- ability to reference canonical names of other types in the docstring, such as `Enum.Variant` or `Error.AccessDenied`. This is important, because enums and errors are represented differently in different languages.

The PR made by @zduny solves the first 3 of these points, but I think the last 2 points are also important to provide the best user experience(function error values and referencing canonical enum/error types in the docstring).

At the moment, due to lack of clarity and an urgent need to improve user experience for our bindings consumers, we have decided that plain docstring text is enough, i.e. copy-paste the text from docstring without attempting to satisfy the above requirements. We are probably going to maintain our own fork with plain-text docstrings, until some form of docstrings are merged to `mozilla/uniffi-rs`.

**Comment by bendk** [comment]
> At the moment, due to lack of clarity and an urgent need to improve user experience for our bindings consumers, we have decided that plain docstring text is enough, i.e. copy-paste the text from docstring without attempting to satisfy the above requirements. We are probably going to maintain our own fork with plain-text docstrings, until some form of docstrings are merged to `mozilla/uniffi-rs`.

If something like the other PR was merged, could you hook up your UDL-based doc parser to it?  It seems like it should be possible, but with some impedance mismatch.  For example, maybe you would only fill out the `description` field of [Structure](https://github.com/mozilla/uniffi-rs/pull/1498/files#diff-8473a4764c13782d4e1fb5d2c13aa0407f391b89e02974b72671da6fe1dff0d5R104-R114) and leave `members` and `methods` empty.

**Comment by arg0d** [comment]
For sure I could hook up UDL docstring parsing to the other PR's code. The structural changes in `interface/*` nodes (function, object, etc..) are basically the same, except that in my PR I'm storing `Option<String>`, while the other PR is storing `Option<uniffi_doc::*>`. Inside the UDL convert functions, I could just feed UDL docstring into doc structures from the other PR, and the result would be identical docstring formats for both Rust and UDL docstrings.

**Comment by arg0d** [comment]
I have updated this PR to include the latest changes from our fork. The implementation is quite straightforward - what you type in UDL is what you get in bindings. Docstrings are lines prefixed with `///`. Docstrings can be prefixed to any element in UDL. The docstring data flows quite naturally from weedle, into metadata, into component interface. Because there are no transformation applied to docstrings, so there isn't much conginitive load added when dealing with the code. Docstrings are included in Kotlin, Swift, and Python bindings. Unit tests are included to validate docstrings at component interface level. Fixture tests are included to validate that docstrings are included in the templates.

File: uniffi_bindgen/src/interface/mod.rs
Comment (arg0d) [inline uniffi_bindgen/src/interface/mod.rs:133]: This is the only sketchy thing in this PR. I didn't want to manually implement the required Eq, PartialEq traits, and this seemed to be the only place where Eq is used.

File: uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceImpl.kt
Comment (mhammond) [inline uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceImpl.kt:3]: is this removal necessary?

**Review by mhammond — CHANGES_REQUESTED** [review]
Thanks, this is exciting! I've a few nits and other suggestions but in general this is looking to be in good shape.

File: fixtures/docstring/tests/bindings/test_docstring.py
Comment (arg0d) [reply fixtures/docstring/tests/bindings/test_docstring.py:11]: I added checking against specific values, but I wasn't sure how I should add checking for none. Should I just create another set of types without docstrings?

**Review by arg0d — COMMENTED** [review]
I hope I dealt/replied with all the comments. Do you guys resolve the trivial comment yourself, or should I resolve the trivial comments after fixing them?

File: uniffi_meta/src/lib.rs
Comment (arg0d) [reply uniffi_meta/src/lib.rs]: I have moved namespace docstring into `MetadataGroup`.

**Comment by mhammond** [comment]
> Do you guys resolve the trivial comment yourself, or should I resolve the trivial comments after fixing them?

We don't really mind - it's fine for you to do it yourself if it helps remember what's still outstanding.

File: fixtures/docstring/tests/bindings/test_docstring.kts
Comment (mhammond) [reply fixtures/docstring/tests/bindings/test_docstring.kts:5]: TBH I think a better option here might be to sprinkle a few docstrings in another fixture (eg, coveralls), with that other fixture *not* checking the docstring semantics, with this fixture trying to limit itself to just the docstring semantics. If every fixture had to test it's main point for existing *and* the union of all other capabilities just to ensure it didn't break something else things would be unmanageable quickly.

Eg, in this particular use-case, I think it sends a better signal to *not* have the test here at all (or a very very small skeleton) so the casual reader doesn't misunderstand what is being tested here.

(It's a shame it's not trivial to extract these, but I see no reason to block landing on a solution to this. It might be worth getting an issue on file after we land to help motivate some future contributor)

File: fixtures/docstring/tests/bindings/test_docstring.py
Comment (mhammond) [reply fixtures/docstring/tests/bindings/test_docstring.py:11]: I think exactly 1 type without a docstring, then check `__doc__ is None` for that type. When first reading this patch I thought that was going to fail - I then saw that it would work, but to help guard against future refactoring it seems worthwhile.

File: fixtures/docstring/tests/test_generated_bindings.rs
Comment (mhammond) [reply fixtures/docstring/tests/test_generated_bindings.rs:14]: I'm suggesting that if I had to choose between, say, `<docstring-associated-enum-variant>` being in a large set vs an explicit test that `Enum1.Foo.__doc__ == "<docstring-associated-enum-variant>"`, the latter has more value IMO. If the (say) Python test checked *every* one of these docstrings in the correct context (ie, by examining `__doc__` for the type it is associated with), we have *both* a check of the entire set *and* ensure that, at least for Python, the docstrings were correctly associated with the type they are meant for.

File: uniffi_bindgen/src/bindings/python/templates/EnumTemplate.py
Comment (mhammond) [reply uniffi_bindgen/src/bindings/python/templates/EnumTemplate.py:26]: Given this is purely stylistic I'm not too bothered by this, but I think the example you pasted is quite fine and isn't going to mislead. I just had a quick look at the python stdlib and the very first docstring I found was https://github.com/python/cpython/blob/main/Lib/calendar.py#L141-L143 and has no blank lines. (But if you scroll a little further though you will find them, so 🤷 .)

If this is intentional I'm fine with it 👍

File: fixtures/docstring/tests/bindings/test_docstring.kts
Comment (arg0d) [reply fixtures/docstring/tests/bindings/test_docstring.kts:5]: I'm not exactly sure what is confusing you in this test. Just like the other tests, this test ensures that bindings code is generated correctly when docstrings are attached. This test *is* limiting itself to docstrings semantics. The test is not actually testing the functionality of other feature, merely the presence of them. Callback is not being invoked, errors are not being thrown. I don't see any other way to test that docstrings aren't breaking the syntax of generated bindings. I could add this docstrings to `coverall`, but I'm not sure I agree that making mega fixtures like coverall is the way to go. Overtime they will just continue to grow, and become more and more unmanagable.

File: fixtures/docstring/tests/test_generated_bindings.rs
Comment (badboy) [reply fixtures/docstring/tests/test_generated_bindings.rs:14]: Can we file a bug for that?

**Review by badboy — APPROVED** [review]
I have little to add after Mark's review, except 👏 and thanks for the extensive work here.
I tested it briefly on our code base and it does exactly what it needs to do.

We should however tackle some documentation about this in a followup and call this out in the changelog.

**Comment by arg0d** [comment]
Thanks @badboy. 

I'm still waiting for feedback/resolution of the remaining conversations. I will delay rebasing on main and fixing conflicts until remaining conversations are resolved to avoid mangling the ongoing conversations.

**Review by mhammond — APPROVED** [review]
I'm perfectly fine with my conversations here being considered resolved and look forward to being able to use this feature! Was there any other specific conversations you wanted to see resolved here?

**Comment by arg0d** [comment]
Oh, ok, I thought you would want to close the "more important" conversations yourself. In that case I will close the remaining converstations myself.

File: fixtures/docstring/tests/test_generated_bindings.rs
Comment (arg0d) [reply fixtures/docstring/tests/test_generated_bindings.rs:14]: Created #1838

**Comment by arg0d** [comment]
Any updates on this? Thoughts, concerns?

**Comment by mhammond** [comment]
Sorry for the delay - I was just about to hit "merge" and noticed the conflicts :( While fixing them up, please feel free to add a CHANGELOG entry, otherwise please open another issue to add it and docs.

**Comment by badboy** [comment]
Ah yes. In the meantime I reviewed the other PR as well and actually wanted to bring it up in chat or with other uniffi-maintainers but didn't yet get a chance to.

This one can definitely land once we have the merge conflicts resolved.

**Comment by arg0d** [comment]
Should I rebase on main, or merge into this branch?

**Comment by mhammond** [comment]
Rebase is probably easiest, but regardless, it will probably be merged with "squash and merge" - ie, we'll probably try and avoid a merge commit in either case

**Comment by arg0d** [comment]
I have rebased and squashed the branch, [added some documentation](https://github.com/mozilla/uniffi-rs/pull/1493/commits/a1aa102eaeea8bfa1c7aad6fa8b4d9d712551da4), and appended changelog.

**Pull request merged**
